internal/credential/state.go

```diff
@@ -269,29 +281,22 @@ func AwsCredentialPersistedStateFromProto(secrets *structpb.Struct, attrs *Crede
 	return s, nil
 }
 
-// GetCredentialType returns the credential type based on the given AWS AccessKey.
-//
-// https://docs.aws.amazon.com/IAM/latest/UserGuide/security-creds.html#sec-access-keys-and-secret-access-keys
-// Access key IDs beginning with ASIA are temporary credentials access keys that you create using AWS STS operations.
-//
-// https://docs.aws.amazon.com/IAM/latest/UserGuide/security-creds.html#sec-access-keys-and-secret-access-keys
-// Access key IDs beginning with AKIA are long-term access keys for an IAM user or an AWS account root user.
-func GetCredentialType(accessKey string) CredentialType {
-	if strings.HasPrefix(accessKey, "ASIA") {
-		return Dynamic
-	}
-	if strings.HasPrefix(accessKey, "AKIA") {
-		return Static
+// GetCredentialType returns the credential type based on the given
+// AccessKey/RoleARN. See CredentialType definition for more information.
+func GetCredentialType(cc *awsutilv2.CredentialsConfig) CredentialType {
+	if cc == nil {
+		return Unknown
 	}
-	return Unknown
-}
 
-// HasDynamicCredentials returns true if the access key is a dynamic credential type.
-func HasDynamicCredentials(accessKey string) bool {
-	return strings.HasPrefix(accessKey, "ASIA")
-}
+	if len(cc.RoleARN) > 0 || strings.HasPrefix(cc.AccessKey, "ASIA") {
+		return DynamicAWS
+	}
+	if len(cc.AccessKey) > 0 && strings.HasPrefix(cc.AccessKey, "AKIA") {
```

Comment:
Same idea for the static credentials. I feel like the prefix check should cover any scenario.